Software/Development/SemesterProject/App/pushbutttonManager.c

```diff
@@ -0,0 +1,114 @@
+ *      Its goal is to qualify whether a pushbutton is on or not.
+ *      It is based on a FSM which will qualify a 'pressed' event
+ *      whenever a pushbutton is pushed AND released. The pressed event
+ *      will hold a '1' value until an external software component
+ *      reads it. At that moment, the value will be reset to 0.
+ *      This implies risks in case that two software components use the same
+ *      button, but that will never be allowed.
+ *
+ */
+
+#include "pushButtonManager.h"
+
+static struct pushButtonListTag
+{
+    PushButton regenerativeBraking;
+    PushButton torqueReference;
+    PushButton cruiseControl;
+    PushButton antiSlip;
+}pushButtonList;
+
+void initPushbuttons()
+{
+    PushButton *structPointer;
+    PushButton *initialMemoryPosition = &pushButtonList.regenerativeBraking;
+    PushButton *finalMemoryPosition = initialMemoryPosition + sizeof(pushButtonList)/sizeof(PushButton);
+
+    pushButtonList.regenerativeBraking.functionPointer = isRegenBrakingPushbuttonEnabled;
+    pushButtonList.antiSlip.functionPointer = isAntiSlipPushbuttonEnabled;
+    pushButtonList.cruiseControl.functionPointer = isCruiseControlPushbuttonEnabled;
+    pushButtonList.torqueReference.functionPointer = isTorqueReferencePushButtonEnabled;
+
+    for (structPointer = initialMemoryPosition; structPointer < finalMemoryPosition; structPointer++)
+    {
+        structPointer->FSMState = waitingForRisingEdge; // First State of FSM
+        structPointer->ButtonState = 0;
+    }
+}
+
+void handlePushbuttons(void)
+{
+    PushButton *structPointer;
+    PushButton *initialMemoryPosition = &pushButtonList.regenerativeBraking;
+    PushButton *finalMemoryPosition = initialMemoryPosition + sizeof(pushButtonList)/sizeof(PushButton);
+
+    for (structPointer = initialMemoryPosition; structPointer < finalMemoryPosition; structPointer++)
+    {
+        pushEventQualificationFSM(structPointer);
+    }
+}
+
+void pushEventQualificationFSM(PushButton *pushbutton)
+{
+    switch(pushbutton->FSMState)
+    {
+        case waitingForRisingEdge:
+        {
+            if (pinIsHigh(pushbutton))  pushbutton->FSMState = waitingForFallingEdge;
+        }
+        case waitingForFallingEdge:
+        {
+            if (pinIsLow(pushbutton))   pushbutton->FSMState = qualifyState;
+        }
+        case qualifyState:
+        {
+            pushbutton->ButtonState = 1;
+            pushbutton->FSMState = waitingForRisingEdge;
+        }
+    }
+}
+
+int pinIsHigh(PushButton *pushbutton)
+{
+    return pushbutton->functionPointer();
+}
+
+int pinIsLow(PushButton *pushbutton)
+{
+    return !(pushbutton->functionPointer());
+}
+
+int regenerativeBrakingHasBeenPressed(void)
+{
+    int buttonStateValue = pushButtonList.regenerativeBraking.ButtonState;
+    pushButtonList.regenerativeBraking.ButtonState = 0;
+    return buttonStateValue;
+}
+
+int cruiseControlHasBeenPressed(void)
```

Comment:
maybe better referenceType?
Also, this function will have to change it's value under circunstances different from button being pressed. For example when the driver brakes while it's active, the reference type has to change from speed to torque. Same for when driver accelerates momentarely or other cases. 
How are this situations going to be handled? maybe better changing the name to referenceTypeHasChanged? 
Maybe this is done somewhere else, idk but its something to keep in mind. Also we have to take a look at priorities and what to do under every situation.

Reply:
reference type ✔️

priorities and doing this kind of situation analysis you describe is performed in reference handler. You are completely right that we definitely have to look very carefully into it.